File: 1000-1999/1003_Rumores.java
import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

public class Main {
    public static void main(String[] arg)
    {
        Scanner in = new Scanner(System.in);
        int t = in.nextInt();
        for (int i = 0; i < t; i++) {
            int n = in.nextInt(); // numero de personas numeradas
            int m = in.nextInt(); // numero de relaciones de amistar
            int matris[][] = new int[2][m];
            for (int j = 0; j < m; j++) {
                int u = in.nextInt();
                int v = in.nextInt();
                matris[0][j]=u;
                matris[1][j]=v;
            }
            int x = in.nextInt();
            int y = in.nextInt();
            funcion(m,x,y,matris);
        }
        
    }

    private static void funcion(int m, int x, int y, int[][] matris) {
        List lista = new ArrayList();
        for (int j = 0; j < m; j++) {
            for (int i = 0; i < m; i++) {
                if(x==matris[0][i])
                {
                    lista.add(matris[1][i]);
                }
            }
            int num = lista.size();
            x = (int) lista.get(num-1);
        }
        int num = lista.size();
        for (int i = 0; i < num; i++) {
            int var = (int) lista.get(i);
            if(var==y)
            {
                System.out.println("SI");
                return;
            }
        }
        System.out.println("NO");
        return;
    }
}

File: 1000-1999/1003_Rumores.cpp
#include <iostream>
#include <vector>

using namespace std;

void funcion(int m, int x, int y, vector<vector<int>> &matris);

int main() {
    int t;
    cin >> t;

    for (int i = 0; i < t; i++) {
        int n, m;
        cin >> n >> m;

        vector<vector<int>> matris(m, vector<int>(2));
        for (int j = 0; j < m; j++) {
            int u, v;
            cin >> u >> v;
            matris[j][0] = u;
            matris[j][1] = v;
        }

        int x, y;
        cin >> x >> y;

        funcion(m, x, y, matris);
    }

    return 0;
}

void funcion(int m, int x, int y, vector<vector<int>> &matris) {
    vector<int> lista;
    for (int j = 0; j < m; j++) {
        for (int i = 0; i < m; i++) {
            if (x == matris[i][0]) {
                lista.push_back(matris[i][1]);
            }
        }
        int num = lista.size();
        x = lista[num - 1];
    }
    int num = lista.size();
    for (int i = 0; i < num; i++) {
        int var = lista[i];
        if (var == y) {
            cout << "SI" << endl;
            return;
        }
    }
    cout << "NO" << endl;
}
